File: workmate-server/src/services/strategy.service.ts
import { PrismaClient } from '@prisma/client'
import { logger } from '../utils/logger'

const prisma = new PrismaClient()

export interface CreateStrategyInput {
  name: string
  releaseType: 'UPGRADE' | 'ROLLBACK'
  targetVersion: string
  downloadUrl: string
  releaseNotes?: string
  platform: string
  minVersion?: string
  totalStages: number
  soakTimeMinutes?: number
  autoPauseErrorRate?: number
  autoPauseEnabled?: boolean
  stages: Array<{
    name: string
    releaseNotes?: string
    rules: Array<{
      ruleType: string
      ruleValue: string
    }>
  }>
}

export async function createStrategy(input: CreateStrategyInput) {
  const strategy = await prisma.$transaction(async (tx) => {
    const st = await tx.upgradeStrategy.create({
      data: {
        name: input.name,
        releaseType: input.releaseType,
        targetVersion: input.targetVersion,
        downloadUrl: input.downloadUrl,
        releaseNotes: input.releaseNotes ?? null,
        platform: input.platform,
        minVersion: input.minVersion ?? null,
        totalStages: input.totalStages,
        soakTimeMinutes: input.soakTimeMinutes ?? null,
        autoPauseErrorRate: input.autoPauseErrorRate ?? null,
        autoPauseEnabled: input.autoPauseEnabled ?? false,
      },
    })

    for (let i = 0; i < input.stages.length; i++) {
      const stageInput = input.stages[i]
      const stage = await tx.upgradeStrategyStage.create({
        data: {
          strategyId: st.id,
          stageOrder: i + 1,
          name: stageInput.name,
          releaseNotes: stageInput.releaseNotes ?? null,
        },
      })

      await tx.upgradeStrategyStageRule.createMany({
        data: stageInput.rules.map((rule) => ({
          stageId: stage.id,
          ruleType: rule.ruleType,
          ruleValue: rule.ruleValue,
        })),
      })
    }

    return st
  })

  return strategy
}

export async function activateStrategy(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
    include: { stages: { orderBy: { stageOrder: 'asc' }, include: { rules: true } } },
  })
  if (!strategy) throw new Error('策略不存在')
  if (strategy.status !== 'DRAFT') throw new Error('只有草稿状态的策略可以启动')

  // 同一平台同一时间只能有一个激活策略
  const existingActive = await prisma.upgradeStrategy.findFirst({
    where: { platform: strategy.platform, status: 'ACTIVE' },
  })
  if (existingActive) {
    throw new Error(`平台 ${strategy.platform} 已有激活策略「${existingActive.name}」（ID: ${existingActive.id}），请先完成或暂停该策略`)
  }

  const now = new Date()
  const firstStage = strategy.stages[0]

  await prisma.upgradeStrategy.update({
    where: { id: strategyId },
    data: { status: 'ACTIVE', currentStage: 1 },
  })

  if (firstStage) {
    await prisma.upgradeStrategyStage.update({
      where: { id: firstStage.id },
      data: { advancedAt: now },
    })

    const rules = firstStage.rules
    if (rules && rules.length > 0) {
      await prisma.upgradeWhitelist.createMany({
        data: rules.map(rule => ({
          sourceStrategyId: strategyId,
          ruleType: rule.ruleType,
          ruleValue: rule.ruleValue,
          targetVersion: strategy.targetVersion,
          platform: strategy.platform,
          isActive: true,
        })),
      })
    }
  }

  return prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
    include: { stages: { include: { rules: true } } },
  })
}

/**
 * 激活策略时同步写入/更新 upgrade_releases 表
 * 供其他需要从 releases 表查询的场景使用
 */
async function syncStrategyToReleases(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
  })
  if (!strategy) return

  await prisma.$transaction(async (tx) => {
    await tx.upgradeRelease.updateMany({
      where: { platform: strategy.platform, isActive: true },
      data: { isActive: false },
    })
    await tx.upgradeRelease.create({
      data: {
        version: strategy.targetVersion,
        releaseType: strategy.releaseType as 'UPGRADE' | 'ROLLBACK',
        releaseNotes: strategy.releaseNotes ?? '',
        downloadUrl: strategy.downloadUrl,
        platform: strategy.platform,
        minVersion: strategy.minVersion,
        isActive: true,
      },
    })
  })
}

export async function advanceStrategyStage(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
    include: { stages: { include: { rules: true }, orderBy: { stageOrder: 'asc' } } },
  })

  if (!strategy) throw new Error('策略不存在')
  if (strategy.status !== 'ACTIVE') throw new Error('只有进行中的策略可以推进')

  const nextStage = strategy.stages.find(
    (s) => s.stageOrder === strategy.currentStage + 1
  )
  if (!nextStage) throw new Error('已到达最终阶段')

  const now = new Date()

  const currentStage = strategy.stages.find(
    (s) => s.stageOrder === strategy.currentStage
  )

  if (currentStage?.advancedAt && strategy.soakTimeMinutes) {
    const elapsed = (now.getTime() - currentStage.advancedAt.getTime()) / 60000
    if (elapsed < strategy.soakTimeMinutes) {
      throw new Error(
        `浸泡时间不足，还需等待 ${Math.ceil(strategy.soakTimeMinutes - elapsed)} 分钟`
      )
    }
  }

  if (strategy.autoPauseEnabled && strategy.autoPauseErrorRate) {
    const errorCount = await getStageErrorCount(
      strategy.platform,
      strategy.targetVersion,
      currentStage?.advancedAt ?? undefined
    )
    const totalRequests = await getStageRequestCount(
      strategy.platform,
      strategy.targetVersion,
      currentStage?.advancedAt ?? undefined
    )

    if (totalRequests > 10) {
      const errorRate = errorCount / totalRequests
      if (errorRate > Number(strategy.autoPauseErrorRate)) {
        await prisma.upgradeStrategy.update({
          where: { id: strategyId },
          data: { status: 'PAUSED' },
        })
        throw new Error(
          `错误率 ${(errorRate * 100).toFixed(1)}% 超过阈值 ${(Number(strategy.autoPauseErrorRate) * 100).toFixed(1)}%，策略已自动暂停`
        )
      }
    }
  }

  const result = await prisma.$transaction(async (tx) => {
    const allRules = strategy.stages
      .filter((s) => s.stageOrder <= nextStage.stageOrder)
      .flatMap((s) => s.rules)

    await tx.upgradeWhitelist.deleteMany({
      where: { sourceStrategyId: strategyId },
    })

    if (allRules.length > 0) {
      await tx.upgradeWhitelist.createMany({
        data: allRules.map((rule) => ({
          sourceStrategyId: strategyId,
          ruleType: rule.ruleType,
          ruleValue: rule.ruleValue,
          targetVersion: strategy.targetVersion,
          platform: strategy.platform,
          isActive: true,
        })),
      })
    }

    await tx.upgradeStrategyStage.update({
      where: { id: nextStage.id },
      data: { advancedAt: now },
    })

    await tx.upgradeStrategy.update({
      where: { id: strategyId },
      data: { currentStage: nextStage.stageOrder },
    })

    return nextStage
  })

  return result
}

async function getStageErrorCount(
  platform: string,
  targetVersion: string,
  since?: Date
): Promise<number> {
  return prisma.observabilityEvent.count({
    where: {
      eventType: 'error',
      clientPlatform: platform,
      clientVersion: targetVersion,
      ...(since ? { createdAt: { gte: since } } : {}),
    },
  })
}

async function getStageRequestCount(
  platform: string,
  targetVersion: string,
  since?: Date
): Promise<number> {
  return prisma.observabilityEvent.count({
    where: {
      eventType: { in: ['chat_question', 'agent_question'] },
      clientPlatform: platform,
      clientVersion: targetVersion,
      ...(since ? { createdAt: { gte: since } } : {}),
    },
  })
}

export async function pauseStrategy(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({ where: { id: strategyId } })
  if (!strategy) throw new Error('策略不存在')
  if (strategy.status !== 'ACTIVE') throw new Error('只有进行中的策略可以暂停')

  await prisma.upgradeWhitelist.deleteMany({
    where: { sourceStrategyId: strategyId },
  })

  return prisma.upgradeStrategy.update({
    where: { id: strategyId },
    data: { status: 'PAUSED' },
  })
}

export async function resumeStrategy(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
    include: {
      stages: { include: { rules: true }, orderBy: { stageOrder: 'asc' } },
    },
  })

  if (!strategy) throw new Error('策略不存在')
  if (strategy.status !== 'PAUSED') throw new Error('只有已暂停的策略可以恢复')

  const allRules = strategy.stages
    .filter((s) => s.stageOrder <= strategy.currentStage)
    .flatMap((s) => s.rules)

  await prisma.$transaction(async (tx) => {
    await tx.upgradeWhitelist.deleteMany({
      where: { sourceStrategyId: strategyId },
    })

    if (allRules.length > 0) {
      await tx.upgradeWhitelist.createMany({
        data: allRules.map((rule) => ({
          sourceStrategyId: strategyId,
          ruleType: rule.ruleType,
          ruleValue: rule.ruleValue,
          targetVersion: strategy.targetVersion,
          platform: strategy.platform,
          isActive: true,
        })),
      })
    }

    await tx.upgradeStrategy.update({
      where: { id: strategyId },
      data: { status: 'ACTIVE' },
    })
  })

  return strategy
}

export async function finishStrategy(strategyId: number) {
  const strategy = await prisma.upgradeStrategy.findUnique({ where: { id: strategyId } })
  if (!strategy) throw new Error('策略不存在')

  await prisma.upgradeWhitelist.deleteMany({
    where: { sourceStrategyId: strategyId },
  })

  return prisma.upgradeStrategy.update({
    where: { id: strategyId },
    data: { status: 'FINISHED' },
  })
}

export async function editStrategyStages(
  strategyId: number,
  stages: Array<{
    name: string
    rules: Array<{ ruleType: string; ruleValue: string }>
  }>,
  totalStages: number,
) {
  const strategy = await prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
  })
  if (!strategy) throw new Error('策略不存在')
  // 允许 DRAFT / ACTIVE / PAUSED 状态编辑
  if (strategy.status === 'FINISHED') throw new Error('已完成的策略不可编辑')

  return prisma.$transaction(async (tx) => {
    // 保留现有阶段的 advancedAt，按 stageOrder 匹配
    const existingStages = await tx.upgradeStrategyStage.findMany({
      where: { strategyId },
      orderBy: { stageOrder: 'asc' },
    })

    // 删除所有旧阶段（Cascade 会删除关联的 rules）
    await tx.upgradeStrategyStage.deleteMany({
      where: { strategyId },
    })

    // 插入新阶段（保留已有阶段的 advancedAt）
    for (let i = 0; i < stages.length; i++) {
      const stageInput = stages[i]
      const existingStage = existingStages.find(es => es.stageOrder === i + 1)
      const stage = await tx.upgradeStrategyStage.create({
        data: {
          strategyId,
          stageOrder: i + 1,
          name: stageInput.name,
          releaseNotes: null,
          advancedAt: existingStage?.advancedAt ?? null,
        },
      })

      await tx.upgradeStrategyStageRule.createMany({
        data: stageInput.rules.map(rule => ({
          stageId: stage.id,
          ruleType: rule.ruleType,
          ruleValue: rule.ruleValue,
        })),
      })
    }

    await tx.upgradeStrategy.update({
      where: { id: strategyId },
      data: { totalStages },
    })

    // 重新同步当前活跃阶段的白名单
    if (strategy.status === 'ACTIVE') {
      const currentStage = stages[strategy.currentStage - 1]
      if (currentStage && currentStage.rules.length > 0) {
        await tx.upgradeWhitelist.deleteMany({
          where: { sourceStrategyId: strategyId },
        })
        await tx.upgradeWhitelist.createMany({
          data: currentStage.rules.map(rule => ({
            sourceStrategyId: strategyId,
            ruleType: rule.ruleType,
            ruleValue: rule.ruleValue,
            targetVersion: strategy.targetVersion,
            platform: strategy.platform,
            isActive: true,
          })),
        })
      }
    }

    return tx.upgradeStrategy.findUnique({
      where: { id: strategyId },
      include: { stages: { include: { rules: true } } },
    })
  })
}

export async function listStrategies(params: { page: number; pageSize: number }) {
  const [total, strategies] = await Promise.all([
    prisma.upgradeStrategy.count(),
    prisma.upgradeStrategy.findMany({
      skip: (params.page - 1) * params.pageSize,
      take: params.pageSize,
      orderBy: { createdAt: 'desc' },
      include: {
        _count: { select: { stages: true } },
      },
    }),
  ])

  return { total, strategies }
}

export async function getStrategyDetail(strategyId: number) {
  return prisma.upgradeStrategy.findUnique({
    where: { id: strategyId },
    include: {
      stages: {
        orderBy: { stageOrder: 'asc' },
        include: { rules: true },
      },
    },
  })
}

File: workmate-server-java/src/main/java/com/workmate/server/service/StrategyService.java
package com.workmate.server.service;

import com.workmate.server.dto.request.StrategyCreateRequest;
import com.workmate.server.dto.response.PaginatedData;
import com.workmate.server.entity.UpgradeStrategy;
import com.workmate.server.entity.UpgradeStrategyStage;
import com.workmate.server.entity.UpgradeStrategyStageRule;
import com.workmate.server.entity.UpgradeWhitelist;
import com.workmate.server.exception.AppException;
import com.workmate.server.mapper.ObservabilityErrorMapper;
import com.workmate.server.mapper.ObservabilityEventMapper;
import com.workmate.server.mapper.UpgradeStrategyMapper;
import com.workmate.server.mapper.UpgradeStrategyStageMapper;
import com.workmate.server.mapper.UpgradeStrategyStageRuleMapper;
import com.workmate.server.mapper.UpgradeWhitelistMapper;
import com.github.pagehelper.PageHelper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class StrategyService {

    private final UpgradeStrategyMapper strategyMapper;
    private final UpgradeStrategyStageMapper stageMapper;
    private final UpgradeStrategyStageRuleMapper stageRuleMapper;
    private final UpgradeWhitelistMapper whitelistMapper;
    private final ObservabilityEventMapper eventMapper;
    private final ObservabilityErrorMapper errorMapper;

    @Transactional
    public UpgradeStrategy createStrategy(StrategyCreateRequest input) {
        UpgradeStrategy strategy = UpgradeStrategy.builder()
                .name(input.getName())
                .releaseType(input.getReleaseType().name())
                .targetVersion(input.getTargetVersion())
                .downloadUrl(input.getDownloadUrl())
                .releaseNotes(input.getReleaseNotes())
                .platform(input.getPlatform())
                .minVersion(input.getMinVersion())
                .totalStages(input.getTotalStages())
                .currentStage(0)
                .soakTimeMinutes(input.getSoakTimeMinutes())
                .autoPauseErrorRate(input.getAutoPauseErrorRate())
                .autoPauseEnabled(input.getAutoPauseEnabled() != null ? input.getAutoPauseEnabled() : false)
                .status("DRAFT")
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        strategyMapper.insert(strategy);

        for (int i = 0; i < input.getStages().size(); i++) {
            StrategyCreateRequest.StageInput stageInput = input.getStages().get(i);
            UpgradeStrategyStage stage = UpgradeStrategyStage.builder()
                    .strategyId(strategy.getId())
                    .stageOrder(i + 1)
                    .name(stageInput.getName())
                    .releaseNotes(stageInput.getReleaseNotes())
                    .createdAt(LocalDateTime.now())
                    .build();
            stageMapper.insert(stage);

            if (stageInput.getRules() != null) {
                for (StrategyCreateRequest.RuleInput ruleInput : stageInput.getRules()) {
                    UpgradeStrategyStageRule rule = UpgradeStrategyStageRule.builder()
                            .stageId(stage.getId())
                            .ruleType(ruleInput.getRuleType())
                            .ruleValue(ruleInput.getRuleValue())
                            .build();
                    stageRuleMapper.insert(rule);
                }
            }
        }

        return strategyMapper.findById(strategy.getId()).orElse(strategy);
    }

    @Transactional
    public UpgradeStrategy activateStrategy(Integer strategyId) {
        UpgradeStrategy strategy = strategyMapper.findById(strategyId)
                .orElseThrow(() -> new AppException(404, "策略不存在"));
        if (!"DRAFT".equals(strategy.getStatus())) {
            throw new AppException(400, "只有草稿状态的策略可以启动");
        }

        strategyMapper.findFirstByPlatformAndStatus(strategy.getPlatform(), "ACTIVE")
                .ifPresent(existing -> {
                    throw new AppException(400, "平台 " + strategy.getPlatform() +
                            " 已有激活策略「" + existing.getName() + "」（ID: " + existing.getId() + "），请先完成或暂停该策略");
                });

        LocalDateTime now = LocalDateTime.now();
        strategy.setStatus("ACTIVE");
        strategy.setCurrentStage(1);
        strategy.setUpdatedAt(now);
        strategyMapper.update(strategy);

        List<UpgradeStrategyStage> stages = stageMapper.findByStrategyIdOrderByStageOrderAsc(strategyId);
        if (!stages.isEmpty()) {
            UpgradeStrategyStage firstStage = stages.get(0);
            firstStage.setAdvancedAt(now);
            stageMapper.update(firstStage);

            List<UpgradeStrategyStageRule> rules = stageRuleMapper.findByStageId(firstStage.getId());
            if (rules != null && !rules.isEmpty()) {
                for (UpgradeStrategyStageRule rule : rules) {
                    UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                            .sourceStrategyId(strategyId)
                            .ruleType(rule.getRuleType())
                            .ruleValue(rule.getRuleValue())
                            .targetVersion(strategy.getTargetVersion())
                            .platform(strategy.getPlatform())
                            .isActive(true)
                            .createdAt(LocalDateTime.now())
                            .build();
                    whitelistMapper.insert(whitelist);
                }
            }
        }

        return strategyMapper.findById(strategyId).orElse(strategy);
    }

    @Transactional
    public UpgradeStrategyStage advanceStrategyStage(Integer strategyId) {
        UpgradeStrategy strategy = findStrategyWithStages(strategyId);
        if (!"ACTIVE".equals(strategy.getStatus())) {
            throw new AppException(400, "只有进行中的策略可以推进");
        }

        List<UpgradeStrategyStage> stages = stageMapper.findByStrategyIdOrderByStageOrderAsc(strategyId);
        UpgradeStrategyStage nextStage = stages.stream()
                .filter(s -> s.getStageOrder() == strategy.getCurrentStage() + 1)
                .findFirst()
                .orElseThrow(() -> new AppException(400, "已到达最终阶段"));

        LocalDateTime now = LocalDateTime.now();

        UpgradeStrategyStage currentStage = stages.stream()
                .filter(s -> s.getStageOrder().equals(strategy.getCurrentStage()))
                .findFirst().orElse(null);

        if (currentStage != null && currentStage.getAdvancedAt() != null && strategy.getSoakTimeMinutes() != null) {
            long elapsedMinutes = java.time.Duration.between(currentStage.getAdvancedAt(), now).toMinutes();
            if (elapsedMinutes < strategy.getSoakTimeMinutes()) {
                throw new AppException(400, "浸泡时间不足，还需等待 " +
                        (strategy.getSoakTimeMinutes() - elapsedMinutes) + " 分钟");
            }
        }

        if (Boolean.TRUE.equals(strategy.getAutoPauseEnabled()) && strategy.getAutoPauseErrorRate() != null) {
            long errorCount = getStageErrorCount(strategy.getPlatform(), strategy.getTargetVersion(),
                    currentStage != null ? currentStage.getAdvancedAt() : null);
            long totalRequests = getStageRequestCount(strategy.getPlatform(), strategy.getTargetVersion(),
                    currentStage != null ? currentStage.getAdvancedAt() : null);

            if (totalRequests > 10) {
                double errorRate = (double) errorCount / totalRequests;
                if (errorRate > strategy.getAutoPauseErrorRate().doubleValue()) {
                    strategy.setStatus("PAUSED");
                    strategy.setUpdatedAt(now);
                    strategyMapper.update(strategy);
                    throw new AppException(400, String.format(
                            "错误率 %.1f%% 超过阈值 %.1f%%，策略已自动暂停",
                            errorRate * 100, strategy.getAutoPauseErrorRate().doubleValue() * 100));
                }
            }
        }

        whitelistMapper.deleteByStrategyId(strategyId);

        List<UpgradeStrategyStageRule> allRules = stages.stream()
                .filter(s -> s.getStageOrder() <= nextStage.getStageOrder())
                .flatMap(s -> stageRuleMapper.findByStageId(s.getId()).stream())
                .toList();

        for (UpgradeStrategyStageRule rule : allRules) {
            UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                    .sourceStrategyId(strategyId)
                    .ruleType(rule.getRuleType()).ruleValue(rule.getRuleValue())
                    .targetVersion(strategy.getTargetVersion()).platform(strategy.getPlatform())
                    .isActive(true).createdAt(LocalDateTime.now())
                    .build();
            whitelistMapper.insert(whitelist);
        }

        nextStage.setAdvancedAt(now);
        stageMapper.update(nextStage);

        strategy.setCurrentStage(nextStage.getStageOrder());
        strategy.setUpdatedAt(now);
        strategyMapper.update(strategy);

        return nextStage;
    }

    @Transactional
    public UpgradeStrategy retreatStrategyStage(Integer strategyId) {
        UpgradeStrategy strategy = findStrategyWithStages(strategyId);
        if (!"ACTIVE".equals(strategy.getStatus())) {
            throw new AppException(400, "只有进行中的策略可以回撤");
        }
        if (strategy.getCurrentStage() <= 1) {
            throw new AppException(400, "已在第一阶段，无法回撤");
        }

        int prevStageOrder = strategy.getCurrentStage() - 1;
        List<UpgradeStrategyStage> stages = stageMapper.findByStrategyIdOrderByStageOrderAsc(strategyId);
        UpgradeStrategyStage prevStage = stages.stream()
                .filter(s -> s.getStageOrder() == prevStageOrder)
                .findFirst()
                .orElseThrow(() -> new AppException(400, "上一阶段不存在"));

        List<UpgradeStrategyStageRule> prevRules = stageRuleMapper.findByStageId(prevStage.getId());
        boolean isPrevFullRollout = prevRules == null || prevRules.isEmpty();

        whitelistMapper.deleteByStrategyId(strategyId);

        if (!isPrevFullRollout) {
            List<UpgradeStrategyStageRule> prevAllRules = stages.stream()
                    .filter(s -> s.getStageOrder() <= prevStageOrder)
                    .flatMap(s -> stageRuleMapper.findByStageId(s.getId()).stream())
                    .toList();

            for (UpgradeStrategyStageRule rule : prevAllRules) {
                UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                        .sourceStrategyId(strategyId)
                        .ruleType(rule.getRuleType()).ruleValue(rule.getRuleValue())
                        .targetVersion(strategy.getTargetVersion()).platform(strategy.getPlatform())
                        .isActive(true).createdAt(LocalDateTime.now())
                        .build();
                whitelistMapper.insert(whitelist);
            }
        }

        strategy.setCurrentStage(prevStageOrder);
        strategy.setUpdatedAt(LocalDateTime.now());
        strategyMapper.update(strategy);

        return strategyMapper.findById(strategyId).orElse(strategy);
    }

    @Transactional
    public UpgradeStrategy pauseStrategy(Integer strategyId) {
        UpgradeStrategy strategy = strategyMapper.findById(strategyId)
                .orElseThrow(() -> new AppException(404, "策略不存在"));
        if (!"ACTIVE".equals(strategy.getStatus())) {
            throw new AppException(400, "只有进行中的策略可以暂停");
        }

        whitelistMapper.deleteByStrategyId(strategyId);
        strategy.setStatus("PAUSED");
        strategy.setUpdatedAt(LocalDateTime.now());
        strategyMapper.update(strategy);
        return strategy;
    }

    @Transactional
    public UpgradeStrategy resumeStrategy(Integer strategyId) {
        UpgradeStrategy strategy = findStrategyWithStages(strategyId);
        if (!"PAUSED".equals(strategy.getStatus())) {
            throw new AppException(400, "只有已暂停的策略可以恢复");
        }

        List<UpgradeStrategyStage> stages = stageMapper.findByStrategyIdOrderByStageOrderAsc(strategyId);
        List<UpgradeStrategyStageRule> allRules = stages.stream()
                .filter(s -> s.getStageOrder() <= strategy.getCurrentStage())
                .flatMap(s -> stageRuleMapper.findByStageId(s.getId()).stream())
                .toList();

        whitelistMapper.deleteByStrategyId(strategyId);

        for (UpgradeStrategyStageRule rule : allRules) {
            UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                    .sourceStrategyId(strategyId)
                    .ruleType(rule.getRuleType()).ruleValue(rule.getRuleValue())
                    .targetVersion(strategy.getTargetVersion()).platform(strategy.getPlatform())
                    .isActive(true).createdAt(LocalDateTime.now())
                    .build();
            whitelistMapper.insert(whitelist);
        }

        strategy.setStatus("ACTIVE");
        strategy.setUpdatedAt(LocalDateTime.now());
        strategyMapper.update(strategy);

        return strategy;
    }

    @Transactional
    public UpgradeStrategy finishStrategy(Integer strategyId, Integer nextStrategyId) {
        UpgradeStrategy strategy = strategyMapper.findById(strategyId)
                .orElseThrow(() -> new AppException(404, "策略不存在"));
        if (!"ACTIVE".equals(strategy.getStatus()) && !"PAUSED".equals(strategy.getStatus())) {
            throw new AppException(400, "只有进行中或已暂停的策略可以完成");
        }
        if (nextStrategyId == null) {
            throw new AppException(400, "必须指定下一个升级策略，以保证升级服务不中断");
        }

        UpgradeStrategy nextStrategy = strategyMapper.findById(nextStrategyId)
                .orElseThrow(() -> new AppException(404, "下一个策略不存在"));
        if (!nextStrategy.getPlatform().equals(strategy.getPlatform())) {
            throw new AppException(400, "下一个策略平台不匹配（当前: " + strategy.getPlatform() +
                    "，目标: " + nextStrategy.getPlatform() + "）");
        }
        if (!"DRAFT".equals(nextStrategy.getStatus())) {
            throw new AppException(400, "下一个策略必须为草稿状态");
        }

        LocalDateTime now = LocalDateTime.now();

        whitelistMapper.deleteByStrategyId(strategyId);
        strategy.setStatus("FINISHED");
        strategy.setUpdatedAt(now);
        strategyMapper.update(strategy);

        nextStrategy.setStatus("ACTIVE");
        nextStrategy.setCurrentStage(1);
        nextStrategy.setUpdatedAt(now);
        strategyMapper.update(nextStrategy);

        List<UpgradeStrategyStage> nextStages = stageMapper.findByStrategyIdOrderByStageOrderAsc(nextStrategyId);
        if (!nextStages.isEmpty()) {
            UpgradeStrategyStage firstStage = nextStages.get(0);
            firstStage.setAdvancedAt(now);
            stageMapper.update(firstStage);

            List<UpgradeStrategyStageRule> rules = stageRuleMapper.findByStageId(firstStage.getId());
            if (rules != null && !rules.isEmpty()) {
                for (UpgradeStrategyStageRule rule : rules) {
                    UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                            .sourceStrategyId(nextStrategyId)
                            .ruleType(rule.getRuleType()).ruleValue(rule.getRuleValue())
                            .targetVersion(nextStrategy.getTargetVersion()).platform(nextStrategy.getPlatform())
                            .isActive(true).createdAt(LocalDateTime.now())
                            .build();
                    whitelistMapper.insert(whitelist);
                }
            }
        }

        return strategyMapper.findById(strategyId).orElse(strategy);
    }

    @Transactional
    public UpgradeStrategy editStrategyStages(Integer strategyId,
                                                List<StrategyCreateRequest.StageInput> stagesInput,
                                                Integer totalStages) {
        UpgradeStrategy strategy = strategyMapper.findById(strategyId)
                .orElseThrow(() -> new AppException(404, "策略不存在"));
        if ("FINISHED".equals(strategy.getStatus())) {
            throw new AppException(400, "已完成的策略不可编辑");
        }

        List<UpgradeStrategyStage> existingStages = stageMapper.findByStrategyIdOrderByStageOrderAsc(strategyId);
        java.util.Map<Integer, LocalDateTime> advancedAtMap = new java.util.HashMap<>();
        for (UpgradeStrategyStage es : existingStages) {
            advancedAtMap.put(es.getStageOrder(), es.getAdvancedAt());
        }

        stageMapper.deleteByStrategyId(strategyId);

        for (int i = 0; i < stagesInput.size(); i++) {
            StrategyCreateRequest.StageInput stageInput = stagesInput.get(i);
            UpgradeStrategyStage stage = UpgradeStrategyStage.builder()
                    .strategyId(strategyId)
                    .stageOrder(i + 1)
                    .name(stageInput.getName())
                    .releaseNotes(stageInput.getReleaseNotes())
                    .advancedAt(advancedAtMap.getOrDefault(i + 1, null))
                    .createdAt(LocalDateTime.now())
                    .build();
            stageMapper.insert(stage);

            if (stageInput.getRules() != null) {
                for (StrategyCreateRequest.RuleInput ruleInput : stageInput.getRules()) {
                    UpgradeStrategyStageRule rule = UpgradeStrategyStageRule.builder()
                            .stageId(stage.getId())
                            .ruleType(ruleInput.getRuleType())
                            .ruleValue(ruleInput.getRuleValue())
                            .build();
                    stageRuleMapper.insert(rule);
                }
            }
        }

        strategy.setTotalStages(totalStages);
        strategy.setUpdatedAt(LocalDateTime.now());
        strategyMapper.update(strategy);

        if ("ACTIVE".equals(strategy.getStatus())) {
            StrategyCreateRequest.StageInput currentStageInput = stagesInput.get(strategy.getCurrentStage() - 1);
            if (currentStageInput != null && currentStageInput.getRules() != null && !currentStageInput.getRules().isEmpty()) {
                whitelistMapper.deleteByStrategyId(strategyId);
                for (StrategyCreateRequest.RuleInput ruleInput : currentStageInput.getRules()) {
                    UpgradeWhitelist whitelist = UpgradeWhitelist.builder()
                            .sourceStrategyId(strategyId)
                            .ruleType(ruleInput.getRuleType()).ruleValue(ruleInput.getRuleValue())
                            .targetVersion(strategy.getTargetVersion()).platform(strategy.getPlatform())
                            .isActive(true).createdAt(LocalDateTime.now())
                            .build();
                    whitelistMapper.insert(whitelist);
                }
            }
        }

        return strategyMapper.findById(strategyId).orElse(strategy);
    }

    @Transactional(readOnly = true)
    public PaginatedData<List<UpgradeStrategy>> listStrategies(int page, int pageSize) {
        PageHelper.startPage(page, pageSize);
        List<UpgradeStrategy> list = strategyMapper.findAll();
        long total = strategyMapper.count();
        return PaginatedData.of(list, total, page, pageSize);
    }

    @Transactional(readOnly = true)
    public UpgradeStrategy getStrategyDetail(Integer strategyId) {
        return strategyMapper.findById(strategyId).orElse(null);
    }

    private UpgradeStrategy findStrategyWithStages(Integer strategyId) {
        return strategyMapper.findById(strategyId)
                .orElseThrow(() -> new AppException(404, "策略不存在"));
    }

    private long getStageErrorCount(String platform, String targetVersion, LocalDateTime since) {
        if (since != null) {
            return errorMapper.countByClientPlatformAndClientVersionAndCreatedAtGreaterThanEqual(
                    platform, targetVersion, since);
        }
        return errorMapper.countByClientPlatformAndClientVersionAndCreatedAtGreaterThanEqual(
                platform, targetVersion, LocalDateTime.of(2000, 1, 1, 0, 0));
    }

    private long getStageRequestCount(String platform, String targetVersion, LocalDateTime since) {
        List<String> eventTypes = List.of("chat_question", "agent_question");
        if (since != null) {
            return eventMapper.countByClientPlatformAndClientVersionAndCreatedAtGreaterThanEqualAndEventTypeIn(
                    platform, targetVersion, since, eventTypes);
        }
        return eventMapper.countByClientPlatformAndClientVersionAndCreatedAtGreaterThanEqualAndEventTypeIn(
                platform, targetVersion, LocalDateTime.of(2000, 1, 1, 0, 0), eventTypes);
    }
}
